main/ptrarray.c

```diff
@@ -0,0 +1,116 @@
+/*
+*   Copyright (c) 1999-2002, Darren Hiebert
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License version 2 or (at your option) any later version.
+*
+*   This module contains functions managing resizable pointer arrays.
+*/
+
+/*
+*   INCLUDE FILES
+*/
+#include "general.h"  /* must always come first */
+
+#include <string.h>
+
+#include "debug.h"
+#include "ptrarray.h"
+
+/*
+*   FUNCTION DEFINITIONS
+*/
+
+extern ptrArray *ptrArrayNew (ptrArrayDeleteFunc deleteFunc)
+{
+	enum { initialSize = 8 };
+	ptrArray* const result = xMalloc (1, ptrArray);
+	result->max = initialSize;
+	result->count = 0;
+	result->array = xMalloc (initialSize, void*);
+	result->deleteFunc = deleteFunc;
+	return result;
+}
+
+extern void ptrArrayAdd (ptrArray *const current, void *ptr)
+{
+	Assert (current != NULL);
+	if (current->count == current->max)
+	{
+		current->max *= 2;
+		current->array = xRealloc (current->array, current->max, void*);
+	}
+	current->array [current->count++] = ptr;
+}
+
+/* Combine array `from' into `current', deleting `from' */
+extern void ptrArrayCombine (ptrArray *const current, ptrArray *const from)
+{
+	unsigned int i;
+	Assert (current != NULL);
+	Assert (from != NULL);
+	for (i = 0  ;  i < from->count  ;  ++i)
+		ptrArrayAdd (current, from->array [i]);
```

Comment:
Well, one would have to make sure the size of dest is enough which ptrArrayAdd() does for us and I'm a bit lazy to do that ;-). Especially when this function doesn't appear in the profiles.